### ISSUE TITLE
[clang][analyzer] False-positive core.NullDereference on an infeasible path

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#222324]
**Description:**
The analyzer loses the proven value of x during interprocedural analysis.

**Reproducer:**
```c
void clang_analyzer_checkInlined(int);
void clang_analyzer_dump(int);

static int f(int);
static int g(int x) { return f(x); }

static int f(int x)
{
    if (x == 2)
        return g(1);

    clang_analyzer_checkInlined(x == 1); // TRUE and FALSE
    clang_analyzer_dump(x); // 1 S32b and reg_$1

    return x || *(int *)0; // core.NullDereference
}

int entry(void)
{
    return f(1);
}
```
Possibly related to #58621

**Comment by NagyDonat** [comment]
This code example was very surprising at first, but I was able to understand it with the help of some debug checkers 😅 

The main trick is that in addition to the obvious `entry()`, the function `g()` also acts as an entrypoint and the `NullDereference` report is emitted during the analysis of `g()`. I noticed this by enabling the `debug.Stats` checker which dumps some statistics _for each entrypoint_ – and in particular reveals each entrypoint [1]. (You can also check that the report still appears if you delete `entry()`; but if you add `int entry2(void) {return g(1);}` the NullDereference disappears because `g` is no longer an entrypoint.)

In this code example the analyzer does the following:
- it selects `entry` as an entrypoint 
  - inlines the call `f(1)`,
  - does enter the `if (x == 2)` conditional because `x` is 1
  - runs through the debug functions
  - returns 1, and does not trigger the null dereference
- then it _also_ selects `g` as an entrypoint, because `g` was not inlined during the analysis of `entry`
  - inlines the call `f(x)` with an unknown symbolic `x`
  - splits the execution path on the `if (x == 2)`
    - one boring path early returns with `g(1)` which eventually yields 1
    - the other continues with the knowledge that `x` is not 2
  - on the continuing path the expression `x == 1` triggers an eagerly assume step, which separates two paths
    - one where x is equal to 1
    - one where x is still unknown (but not 1 and not 2)
  - the `x == 1` path is boring: produces TRUE, then `1 S32b`, then returns `1`
  - the path where `x` is still unknown also reaches the debug functions, producing FALSE, then `reg_$1`
  - this path reaches the short-circuiting `||` which again separates two paths:
    - one where `x` is nonzero, and returned
    - one where `x` is zero, the RHS is evaluated and the analyzer finally sees the null dereference.

Note that multiple identical warnings are deduplicated, this is why you don't see more than one copy of the `debug.ExprInspection` warnings that are emitted on multiple paths.

Based on this, I claim that **this is a true positive report** – the analyzer finds a real null dereference, which is reachable through a feasible path from one of the analyzed functions.

[1] See this [Godbolt reproducer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:16,positionColumn:2,positionLineNumber:16,selectionStartColumn:2,selectionStartLineNumber:16,startColumn:2,startLineNumber:16),source:'void+clang_analyzer_checkInlined(int)%3B%0Avoid+clang_analyzer_dump(int)%3B%0A%0Astatic+int+f(int)%3B%0Astatic+int+g(int+x)+%7B+return+f(x)%3B+%7D%0A%0Astatic+int+f(int+x)%0A%7B%0A++++if+(x+%3D%3D+2)%0A++++++++return+g(1)%3B%0A%0A++++clang_analyzer_checkInlined(x+%3D%3D+1)%3B+//+TRUE+and+FALSE%0A++++clang_analyzer_dump(x)%3B+//+1+S32b+and+reg_$1%0A%0A++++return+x+%7C%7C+*(int+*)0%3B+//+core.NullDereference%0A%7D%0A%0Aint+entry(void)%0A%7B%0A++++return+f(1)%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:41.734417344173444,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'1',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'--analyze+-Xclang+-analyzer-checker+-Xclang+debug.ExprInspection,debug.Stats',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+16.1',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:58.265582655826556,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4); the output of debug.Stats appears as `<function name> -> TotalCFGBlocks: ....`.
